c_glib/arrow-glib/reader.cpp

```diff
@@ -1497,6 +1488,387 @@ garrow_csv_reader_read(GArrowCSVReader *reader,
   }
 }
 
+
+typedef struct GArrowJSONReadOptionsPrivate_ {
+  arrow::MemoryPool *pool;
+  arrow::json::ReadOptions read_options;
+  arrow::json::ParseOptions parse_options;
+  GArrowSchema *schema;
+} GArrowJSONReadOptionsPrivate;
+
+enum {
+  PROP_JSON_READER_USE_THREADS = 1,
+  PROP_JSON_READER_BLOCK_SIZE,
+  PROP_JSON_READER_ALLOW_NEWLINES_IN_VALUES,
+  PROP_JSON_READER_UNEXPECTED_FIELD_BEHAVIOR,
+  PROP_JSON_READER_SCHEMA
+};
+
+G_DEFINE_TYPE_WITH_PRIVATE(GArrowJSONReadOptions,
+                           garrow_json_read_options,
+                           G_TYPE_OBJECT)
+
+#define GARROW_JSON_READ_OPTIONS_GET_PRIVATE(object) \
+  static_cast<GArrowJSONReadOptionsPrivate *>(       \
+    garrow_json_read_options_get_instance_private(   \
+      GARROW_JSON_READ_OPTIONS(object)))
+
+static void
+garrow_json_read_options_dispose(GObject *object)
+{
+  auto priv = GARROW_JSON_READ_OPTIONS_GET_PRIVATE(object);
+
+  if (priv->schema) {
+    g_object_unref(priv->schema);
+    priv->schema = nullptr;
+  }
+
+  G_OBJECT_CLASS(garrow_json_read_options_parent_class)->dispose(object);
+}
+
+static void
+garrow_json_read_options_set_property(GObject *object,
+                                      guint prop_id,
+                                      const GValue *value,
+                                      GParamSpec *pspec)
+{
+  auto priv = GARROW_JSON_READ_OPTIONS_GET_PRIVATE(object);
+
+  switch (prop_id) {
+  case PROP_JSON_READER_USE_THREADS:
+    priv->read_options.use_threads = g_value_get_boolean(value);
+    break;
+  case PROP_JSON_READER_BLOCK_SIZE:
+    priv->read_options.block_size = g_value_get_int(value);
+    break;
+  case PROP_JSON_READER_ALLOW_NEWLINES_IN_VALUES:
+    priv->parse_options.newlines_in_values = g_value_get_boolean(value);
+    break;
+  case PROP_JSON_READER_UNEXPECTED_FIELD_BEHAVIOR:
+    priv->parse_options.unexpected_field_behavior =
+      static_cast<arrow::json::UnexpectedFieldBehavior>(g_value_get_enum(value));
+    break;
+  case PROP_JSON_READER_SCHEMA:
+    priv->schema = GARROW_SCHEMA(g_value_dup_object(value));
```

Comment:
Could you `g_object_unrf()` when `priv->schema` isn't `NULL`? And `g_value_dup_object()` may return `NULL`.

```cpp
if (priv->schema) {
  g_object_unref(priv->schema);
}
auto schema = g_value_dup_object(value);
if (schema) {
  priv->schema = GARROW_SCHEMA(schema);
  priv->parse_options.explicit_schema = garrow_schema_get_raw(priv->schema);
} else {
  priv->schema = NULL;
  priv->parse_options.explicit_schema = nullptr;
}
```

Reply:
I've fixed them.